data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java

```diff
@@ -72,21 +73,39 @@ public KustoOperationResult execute(String database, String command) throws Data
 
     @Override
     public KustoOperationResult execute(String database, String command, ClientRequestProperties properties) throws DataServiceException, DataClientException {
+        String response = executeForJsonResult(database, command, properties);
+
+        String clusterEndpoint = determineClusterEndpoint(command);
+        try {
+            return new KustoOperationResult(response, clusterEndpoint.endsWith("v2/rest/query") ? "v2" : "v1");
+        } catch (KustoServiceError e) {
+            throw new DataClientException(clusterEndpoint, "Error converting json response to KustoOperationResult:" + e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public String executeForJsonResult(String command) throws DataServiceException, DataClientException {
+        return executeForJsonResult(DEFAULT_DATABASE_NAME, command);
+    }
+
+    @Override
+    public String executeForJsonResult(String database, String command) throws DataServiceException, DataClientException {
+        return executeForJsonResult(database, command, null);
+    }
+
+    @Override
+    public String executeForJsonResult(String database, String command, ClientRequestProperties properties) throws DataServiceException, DataClientException {
         // Argument validation:
         if (StringUtils.isAnyEmpty(database, command)) {
-            throw new IllegalArgumentException("database or command are empty");
+            throw new IllegalArgumentException("Database and/or command is empty");
         }
         Long timeoutMs = properties == null ? null : properties.getTimeoutInMilliSec();
 
-        String clusterEndpoint;
-        if (command.startsWith(ADMIN_COMMANDS_PREFIX)) {
-            clusterEndpoint = String.format("%s/%s/rest/mgmt", clusterUrl, MGMT_ENDPOINT_VERSION);
-            if (timeoutMs == null) {
+        String clusterEndpoint = determineClusterEndpoint(command);
+        if (timeoutMs == null) {
+            if (command.startsWith(ADMIN_COMMANDS_PREFIX)) {
```

Comment:
One more thing, but I don't have a good solution for this.

It's a great thing that you exported the logic to determineClusterEndpoint, but then you still have to call command.startsWith(ADMIN_COMMANDS_PREFIX) in your code, which makes it a bit redundant

Reply:
You can probably see I tried to get rid of that redundancy, but ran into the same issue you're calling out.

The root cause is that Java on principle demands that a method return 1 thing. Therefore if the same logic affects multiple parameters, we'd either create a new object or, if the logic is non-expensive, repeat the logic. I chose the latter.

We could add a 1-line method called, "isAdminCommand()" that obscures this repeated line, but all that does is add (little, IMHO) clarity. The logic would still be duplicated.

WDYT?

Reply:
I meant having an isAdminCommand method